### ISSUE TITLE
Rotate SDO maps

### ISSUE THREAD

**Issue by afbattaglia** [samaloney/stixpy#3]
The purpose of this issue is to create a routine in which the user can rotate the SDO AIA (or HMI) maps as it would be seen from Solar Orbiter. This is useful since it may give the possibility to over-plot the STIX images on AIA (or HMI) maps. Indeed the output of the present code should be a SDO map in which the Helioprojective coordinates as seen from Solar Orbiter are given. I think that this can be easily extended to other Earth orbiting (or even Earth based) observatories.

For the moment, I just modified the Jupyter script you provided to Säm a while ago, in a way that at the beginning, under 'Input parameters', one can find and modify directly the parameters to give to the algorithm. These parameters are explained in the Jupyter. Moreover, I added the possibility to give the coordinate and field of view of the flare (or other event) as seen from Earth and then obtain the same map, but as seen from Solar Orbiter. 

So far, I didn't create a real function but I just have run the notebook every time I needed a map. Indeed, before doing this I wanted to know how you intend to organise the code.

However, below I list some ideas that can be added/modified to the routine (among other useful things that certainly exist):
- it may be interesting to give to the user the possibility to visualise the current situation, i.e. plotting the latest SDO map as seen from SolO. I tried to do that with a keyword 'latest' but I'm having troubles of finding the latest available SDO image using Fido;
- it may be also interesting to set a date in the future in which (for example) the latest SDO image is plotted as seen from SolO in the desired future date. This will give an idea of the overlapping observing region on the Sun between Solar Orbiter and Earth based observatories (or Earth orbiting spacecrafts). I wanted to do this since I'm planning a joint observation with some Earth based observatories but unfortunately without success. It seems that in the 'reproject_adaptive' routine, when rotating the map, an error occurs due to the different date of the SDO map and the SolO reference coordinate;
- the 'reproject_adaptive' routine seems to require a considerable amount of memory when rotating the SDO map. To be able to obtain the final map, I had to decrease the size of the output field of view ('out_shape'), since my computer (with 16GB memory) is not able to handle the entire field of view. I tried to reduce the size of the SDO map but the problem is still unresolved. Do you think this is a problem when creating the routine for stixpy? How could we handle this?

**Comment by samaloney** [comment]
This great! So for code layout probably follow sunpy/astropy repo where this works. So the high level function could go under `stixpy/visualisation`  and a stixmap would go under `stixpy/map/STIXMap`

Since as you say it would be useful to be able to do this for many types of observation STEREO, SUIV, SDO, .... it would be good to make it work for all sunpy Maps with the required information.

>it may be interesting to give to the user the possibility to visualise the current situation, i.e. plotting the latest SDO map as seen from SolO. I tried to do that with a keyword 'latest' but I'm having troubles of finding the latest available SDO image using Fido;

I think we should build lower level functions that are more general and then we can built specific functions to plot the latest SDO data as seen from SOLO for example. What the issue with Fido are you getting and error message or just not finding the data, also what version of Sunpy?

>it may be also interesting to set a date in the future in which (for example) the latest SDO image is plotted as seen from SolO in the desired future date. This will give an idea of the overlapping observing region on the Sun between Solar Orbiter and Earth based observatories (or Earth orbiting spacecrafts). I wanted to do this since I'm planning a joint observation with some Earth based observatories but unfortunately without success. It seems that in the 'reproject_adaptive' routine, when rotating the map, an error occurs due to the different date of the SDO map and the SolO reference coordinate;

I'd need to check but I think you would need to differentially rotate the SDO map the the Solo "date" and then reproject the differentially rotated map?

> the 'reproject_adaptive' routine seems to require a considerable amount of memory when rotating the SDO map. To be able to obtain the final map, I had to decrease the size of the output field of view ('out_shape'), since my computer (with 16GB memory) is not able to handle the entire field of view. I tried to reduce the size of the SDO map but the problem is still unresolved. Do you think this is a problem when creating the routine for stixpy? How could we handle this?

Interesting works on my machine with 16 GB, I wonder why it doesn't for you? I think it should be easy to support down sampled image, sunpy can resample/rebin map


I think we need to 
1. Identify the information needed and source e.g. Solo location, STIX FOV, from spice kernels, 
    SDO position from SDO map ...
2. Identify the existing function that we can use in SunPy
3. Figure out what function we need to create and if they below in here or STIXCore, or more general ones in SunPy

Note we will be building some necessary functions over in STIXCore for example the function to get the spacecraft position based on the date, STIX FoV etc  https://github.com/i4Ds/STIXCore

A first cut at function outline could be something like below?

```python

def as_seen_by_SOLO(map, date=None): # this is a terrible name but I can't think of anything else
    """
    Return the input map as seen by SolarOrbiter

    Parameters
    ------------
    map : `sunpy.map.Map`
        Input map to convert to as 
    date : optional `astropy.time.Time` or `datetime.datetime`
        Date  used to obtain SOLO position, defaults to the time of the input map
    """
```

**Comment by afbattaglia** [comment]
Thanks for the reply and information Shane!

> I think we should build lower level functions that are more general and then we can built specific functions to plot the latest SDO data as seen from SOLO for example. What the issue with Fido are you getting and error message or just not finding the data, also what version of Sunpy?

Yes, you are right, that's a really good idea. My Sunpy version is 2.0.3 but finally I found where my error was. It was wrong the way how I was looking up the latest available data. But as you said, maybe now this is not important, we can think about this in a later stage.

> I'd need to check but I think you would need to differentially rotate the SDO map the the Solo "date" and then reproject the differentially rotated map?

Probably yes, I need to better investigate this. Thanks.

> Interesting works on my machine with 16 GB, I wonder why it doesn't for you? I think it should be easy to support down sampled image, sunpy can resample/rebin map

This is indeed really strange. I don't understand why. However, I think this is a problem that has to be taken into account. I will dedicate some time to understand this.

> I think we need to
> 
> 1. Identify the information needed and source e.g. Solo location, STIX FOV, from spice kernels,
>    SDO position from SDO map ...
> 2. Identify the existing function that we can use in SunPy
> 3. Figure out what function we need to create and if they below in here or STIXCore, or more general ones in SunPy
> 
> Note we will be building some necessary functions over in STIXCore for example the function to get the spacecraft position based on the date, STIX FoV etc https://github.com/i4Ds/STIXCore
> 
> A first cut at function outline could be something like below?
> 
> ```python
> def as_seen_by_SOLO(map, date=None): # this is a terrible name but I can't think of anything else
>     """
>     Return the input map as seen by SolarOrbiter
> 
>     Parameters
>     ------------
>     map : `sunpy.map.Map`
>         Input map to convert to as 
>     date : optional `astropy.time.Time` or `datetime.datetime`
>         Date  used to obtain SOLO position, defaults to the time of the input map
>     """
> ```

Ok Shane, thanks a lot for all the info. I will have a look to the three points you listed above and I will let you know as soon as I can.

**Comment by afbattaglia** [comment]
Here I'm again!
Attached to this comment, you can find the python script containing the 'as_seen_by_SOLO' function. Below, some comments.

**How does it work**
The function returns the map you give but as seen by Solar Orbiter and to be able to use it, you need to give the following:
- full disk map. Currently, it works only with FULL DISK map of the SDO/AIA instrument (by chance it may work also for other instruments, but since I'm not an expert on that, I don't know). It doesn't work if you give only a cutout of the full disk map: do we need to include this option?
- the path to the folder containing the SPICE kernels. I don't know how to handle this and most probably the way I did it is not the most useful/efficient way. Any idea? I think this is a general problem that can be included into STIXCore functions... 
- OPTIONAL: date for obtaining the Solar Orbiter position. If it is not given, the date of the full disk map is used.
- OPTIONAL: center. Here you can give the helioprojective coordinates of the center of a region of interest (e.g. flare) on the Solar disc as seen from Earth. If you specify this, the output will be: the full disk map as seen from SolO AND the sub-map of the region of interest as seen from SolO (useful to get the precise coordinates of the region of interest as seen from SolO).
- OPTIONAL: field of view. If you give the coordinates of a region of interest, you can specify the width and height of the FOV (default: 5 arcmin).

**General comments**
Here, a (non-exhaustive) list of general comments:
- the script includes the function 'coordinates_SOLO' which returns the HEE coordinates of Solar Orbiter. To make it work, one has to give the date and the path where the SPICE kernels are stored (see comment above). This may be a STIXCore function, but of course, it has to be improved quite a lot. Does such a function already exist?
- it contains also the function 'roi_hpc_SOLO', which takes the coordinates of the region of interest (ROI) as seen from Earth (on the surface of the Sun) and transform them to helioprojective coordinates as seen from Solar Orbiter (always on the surface of the Sun). Same question as before: does such a function already exist? If not, it may be included in STIXCore I think.
- I found what was the problem with the memory. In the 'reproject' package, there are different routines in order to perform image reprojection and (as far as I understand) every function requires different amount of memory. With 'reproject_adaptive' I need to reduce the shape of the output data in order to not encounter memory problems. With 'reproject_interp' instead, no memory problems seem to appear. Currently in the code I'm using the latter one but I don't know if it is the best option (there are other two functions commented in the code: one 'reproject_adaptive' and the other 'reproject_exact').

This is still under construction and many modifications certainly have still to be applied. I hope that this could be a useful starting point.

[as_seen_by_SOLO.zip](https://github.com/samaloney/stixpy/files/5567858/as_seen_by_SOLO.zip)

**Comment by samaloney** [comment]
Closed with #14
